File: C++ Implementation/Z-Algorithm.cpp
#include<iostream>
#include<fstream>
using namespace std;

void getZarr(string str, int Z[]);

void search(string text, string pattern)
{
 string concat = pattern + "$" + text;
 int l = concat.length();

 int Z[l];
 getZarr(concat, Z);
int count=0;
 for (int i = 0; i < l; ++i)
 {
    if (Z[i] == pattern.length()){
     cout << "Pattern found at index "<< i - pattern.length() -1 << endl;
     count=1;
    }
   }
   if(count==0){
   	cout<<"Pattern not found"<<endl;
   }
}

void getZarr(string str, int Z[])
{
 int n = str.length();
 int L, R, k;

 L = R = 0;
 for (int i = 1; i < n; ++i)
 {
    if (i > R)
    {
       L = R = i;
        while (R<n && str[R-L] == str[R]){
          R++;
         }
       Z[i] = R-L;
       R--;
    }
    else
    {
       k = i-L;
       if (Z[k] < R-i+1){
          Z[i] = Z[k];
        }  
       else {
          L = i;
          while (R<n && str[R-L] == str[R]){
             R++;
          }
          Z[i] = R-L;
          R--;
       }
    }
  }
}

int main()
{	int i;
	    ifstream inputFile("Input-CPP.txt");
	    if (!inputFile) {
	        cerr << "Error: Unable to open input file." << endl;
	        return 1;
	    }
	
		string text,pattern;
		for(i=1;i<=4;i++){
			cout<<"Input "<<i<<":"<<endl;
			if (getline(inputFile, text) && getline(inputFile, pattern)) {
			        search(text, pattern);
			    } 
			else {
			        cerr << "Error: Failed to read input from file." << endl;
			        break;
			    }
	}
	inputFile.close();
	return 0;
}

File: Python Implementation/Z-Algorithm.py
def getZarr(string, z):
    n = len(string)
    l, r, k = 0, 0, 0
    for i in range(1, n):
        if i > r:
            l, r = i, i
            while r < n and string[r - l] == string[r]:
                r += 1
            z[i] = r - l
            r -= 1
        else:
            k = i - l
            if z[k] < r - i + 1:
                z[i] = z[k]
            else:
                l = i
                while r < n and string[r - l] == string[r]:
                    r += 1
                z[i] = r - l
                r -= 1

def search(text, pattern):
    concat = pattern + "$" + text
    l = len(concat)
    z = [0] * l
    getZarr(concat, z)
    occurrences = []
    for i in range(l):
        if z[i] == len(pattern):
            occurrences.append(i - len(pattern) - 1)
    return occurrences

if __name__ == "__main__":
    # Read input from file
    with open("Python Input.txt", "r") as file:
        lines = file.readlines()
        if len(lines) < 2 or len(lines) % 2 != 0:
            print("Error: Input file must contain pairs of text and pattern.")
            exit(1)

        # Process each pair of lines (text and pattern)
        for i in range(0, len(lines), 2):
            text = lines[i].strip()
            pattern = lines[i + 1].strip()

            # Perform pattern search
            results = search(text, pattern)

            # Display results
            if results:
                print(f"Pattern '{pattern}' found at indices:", results)
            else:
                print(f"Pattern '{pattern}' not found in the text '{text}'.")
